lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -314,6 +316,117 @@ class QcFlagService {
+            await getOneDataPassOrFail({ id: dataPassId });
+            await this._validateDataPassAndRunsAssociation(dataPassId, runNumbers);
+            await this._validateRunsAndDplDetectorsAssociations(runNumbers, dplDetectorIds);
+
+            await GaqDetectorRepository.removeAll({
+                where: {
+                    dataPassId,
+                    runNumber: { [Op.in]: runNumbers },
+                },
+            });
+            const gaqEntries = runNumbers
+                .flatMap((runNumber) => dplDetectorIds
+                    .map((dplDetectorId) => ({ dataPassId, runNumber, dplDetectorId })));
+            const createdEntries = await GaqDetectorRepository.insertAll(gaqEntries);
+            return createdEntries.map(gaqDetectorAdapter.toEntity);
+        });
+    }
+
+    /**
+     * Validate association between data pass and runs exists
+     *
+     * @param {number} dataPassId id of a data pass
+     * @param {number[]} runNumbers list of run numbers
+     * @return {Promise<void>} promise resolved once associatios are validated
+     * @throws {BadParameterError} if some association does not exist
+     */
+    async _validateDataPassAndRunsAssociation(dataPassId, runNumbers) {
+        const validRunNumbers = await RunRepository.findAll({
+            attributes: ['runNumber'],
+            where: { runNumber: { [Op.in]: runNumbers } },
+            include: [{ association: 'dataPass', where: { id: dataPassId }, required: false }],
+        });
+        if (validRunNumbers.length !== runNumbers.length) {
+            const validRunNumbersSet = new Set(validRunNumbers.map(({ runNumber }) => runNumber));
+            const invalidRunNumbers = runNumbers.filter((runNumber) => !validRunNumbersSet.has(runNumber));
+            const errorMessage = `No runs for run numbers: (${invalidRunNumbers})`;
+            throw new BadParameterError(errorMessage);
+        }
+        const runsNotAssociatedToDataPass = validRunNumbers
+            .filter(({ dataPass }) => dataPass.length === 0)
+            .map(({ runNumber }) => runNumber);
+        if (runsNotAssociatedToDataPass.length > 0) {
+            const errorMessage = `No association between data pass with id ${dataPassId} and following runs: ${runsNotAssociatedToDataPass}`;
+            throw new BadParameterError(errorMessage);
+        }
+    }
+
+    /**
+     * Validate association between runs and DPL detectors.
+     * As there is no direct link between runs and DPL detectors,
+     *  association between run and corresponding Detector (with same name as DPL Detector) is checked
+     *
+     * @param {number[]} runNumbers list of run numbers
+     * @param {number[]} dplDetectorIds list of IDs of DPL detectors
+     * @return {Promise<void>} promise resolved once associations are validated
+     * @throws {BadParameterError} if some association does not exist
+     */
+    async _validateRunsAndDplDetectorsAssociations(runNumbers, dplDetectorIds) {
```

Comment:
Same here, this can be simplified, if you handle runs one by one

Reply:
https://github.com/AliceO2Group/Bookkeeping/pull/1621#discussion_r1673619269